frontend/src/metabase/visualizations/visualizations/SmartScalar/utils.ts

```diff
@@ -63,3 +79,244 @@ export const getValueHeight = (height: number): number => {
 export const getChangeWidth = (width: number): number => {
   return Math.max(width - ICON_SIZE - ICON_MARGIN_RIGHT - 2 * SPACING, 0);
 };
+
+export const COMPARISON_TYPES = {
+  PREVIOUS_VALUE: "previousValue",
+  PREVIOUS_PERIOD: "previousPeriod",
+  PERIODS_AGO: "periodsAgo",
+} as const;
+
+export type ComparisonTypes =
+  typeof COMPARISON_TYPES[keyof typeof COMPARISON_TYPES];
+
+export const COMPARISON_SELECTOR_OPTIONS = {
+  PREVIOUS_PERIOD: {
+    type: COMPARISON_TYPES.PREVIOUS_PERIOD,
+    prefix: t`Previous`,
+  },
+  PERIODS_AGO: {
+    type: COMPARISON_TYPES.PERIODS_AGO,
+    suffix: t`ago`,
+    MenuItemComponent: PeriodsAgoInputWidget,
+  },
+  PREVIOUS_VALUE: {
+    type: COMPARISON_TYPES.PREVIOUS_VALUE,
+    name: t`Previous value`,
+  },
+};
+
+type PreviousValueMenuOption = {
+  type: typeof COMPARISON_TYPES.PREVIOUS_VALUE;
+  name: string;
+};
+type PreviousPeriodMenuOption = {
+  type: typeof COMPARISON_TYPES.PREVIOUS_PERIOD;
+  name: string;
+};
+type PeriodsAgoMenuOption = {
+  type: typeof COMPARISON_TYPES.PERIODS_AGO;
+  name: string;
+  MenuItemComponent: typeof PeriodsAgoInputWidget;
+  maxValue: number;
+};
+export type ComparisonMenuOption =
+  | PreviousValueMenuOption
+  | PreviousPeriodMenuOption
+  | PeriodsAgoMenuOption;
+
+export type SelectedComparisonPeriodsAgo = {
+  type: typeof COMPARISON_TYPES.PERIODS_AGO;
+  value: number;
+};
+type SelectedComparisonPreviousPeriod = {
+  type: typeof COMPARISON_TYPES.PREVIOUS_PERIOD;
+};
+type SelectedComparisonCompareToPrevious = {
+  type: typeof COMPARISON_TYPES.PREVIOUS_VALUE;
+};
+export type SelectedComparison =
+  | SelectedComparisonCompareToPrevious
+  | SelectedComparisonPreviousPeriod
+  | SelectedComparisonPeriodsAgo;
+
+export function getDefaultComparison(
+  series: RawSeries,
+  settings: VisualizationSettings,
+) {
+  const [
+    {
+      data: { insights },
+    },
+  ] = series;
+
+  const dateUnit = insights?.find(
+    insight => insight.col === settings["scalar.field"],
+  )?.unit as RelativeDatetimeUnit | undefined;
```

Comment:
`insight.unit` has a [`DatetimeUnit` type](https://github.com/metabase/metabase/blob/b4408e033fae30ea08e53037d73ad1dcac45f4dd/frontend/src/metabase-types/api/query.ts#L50) which is way broader than [`RelativeDatetimeUnit`](https://github.com/metabase/metabase/blob/b4408e033fae30ea08e53037d73ad1dcac45f4dd/frontend/src/metabase-types/api/query.ts#L111). IMO type casting here can be dangerous and lead to errors, we should avoid this

Reply:
Thank you for that catch! I actually meant to type `insight.unit` as `RelativeDatetimeUnit`, not `DatetimeUnit`. So, the type casting is not even necessary... :face_in_clouds: